### PR TITLE
printBuffer: Fix stack overflow

### PR DESCRIPTION
**Pull request by EchterAgo**
The `printBuffer` function was not correctly limiting writes to buffer, we passed the full buffer size to `_vsnprintf_s` despite not starting at the first element. This allows for a buffer overflow of 17 bytes after the buffer.

The code is changed to account for this and to truncate the buffer and mark the end with a "TRUNC\n" if the buffer is not large enough.

With the current buffer size, the largest printable string is 1005 bytes (1024 - 18 byte header - 1 byte zero termination). Anything larger will result in the string being truncated.

### CODE REVIEW

**Comment by EchterAgo** [comment]
Updated to fix comment.

**Comment by EchterAgo** [comment]
These prints can be used to help test the change:

```c
// 1005 bytes
printBuffer("%s\n", "notruncAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");

// 1006 bytes
printBuffer("%s\n", "truncAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");
```

The first string should not be truncated while the second one is. I also tested with larger strings and it always produces a buffer with at most 1023 characters and a zero termination byte.

File: module/os/windows/debug.c [hunk 1]
```diff
@@ -125,12 +125,14 @@ printBuffer(const char *fmt, ...)
 	va_list args;
 	va_start(args, fmt);
 	char buf[max_line_length];
-	_snprintf(buf, 18, "%p: ", PsGetCurrentThread());
+	int buf_used;
```
Comment (lundman) [inline module/os/windows/debug.c:128]: I'm not a fan of hardcoded numbers `18` so its a good thing to remove them and do it properly like you have...

File: module/os/windows/debug.c [hunk 2]
```diff
@@ -125,12 +125,14 @@ printBuffer(const char *fmt, ...)
...
 
-	int tmp = _vsnprintf_s(&buf[17], sizeof (buf), max_line_length,
-	    fmt, args);
-	if (tmp >= max_line_length) {
-		_snprintf(&buf[17], 17, "buffer too small");
+	buf_used = _snprintf(buf, sizeof(buf), "%p: ", PsGetCurrentThread());
+
+	int tmp = _vsnprintf_s(buf + buf_used, sizeof(buf) - buf_used, _TRUNCATE, fmt, args);
```
Comment (lundman) [inline module/os/windows/debug.c:132]: ... but you would need to check for `buf_used` being -1 here, or address of `buf - 1` would not go well.
Now there is no real reason why should ever get -1 tho..  `buf` is plenty big for a single %p. Maybe just throw in a `if (buf_used < 0) return;`

**Review by lundman — CHANGES_REQUESTED** [review]
Hey thanks for noticing this and issuing a PR. Sorry about the review nits

File: module/os/windows/debug.c (hunk 2 above)
Comment (EchterAgo) [reply module/os/windows/debug.c:132]: Yea, I was thinking about that too, checking for negative values makes sense even if it should not happen there. I'll update it soon.

Comment (EchterAgo) [reply module/os/windows/debug.c:135]: Updated.

**Comment by lundman** [comment]
Also checkstyle `cstyle.pl` says:
```
./module/os/windows/debug.c: 130: missing space between keyword and paren
Error: missing space between keyword and paren
./module/os/windows/debug.c: 132: line > 80 characters
Error: line > 80 characters
./module/os/windows/debug.c: 132: missing space between keyword and paren
Error: missing space between keyword and paren
```

**Comment by EchterAgo** [comment]
Fixed code style.

**Comment by EchterAgo** [comment]
Note that I'm not sure if there are currently any prints that could hit this issue. I noticed this when adding my own prints to the code.

**Comment by lundman** [comment]
Clearly a bug, even if it doesn't trigger today.

**Review by lundman — APPROVED** [review]

**Comment by EchterAgo** [comment]
```
./module/os/windows/debug.c: 136: continuation should be indented 4 spaces
```

Is the `checkstyle` task using `cstyle.pl`? because I don't see that when running cstyle locally. Will fix.

**Comment by lundman** [comment]
you can run `script/cstyle.pl -cpP` by hand, or they run `make cstyle` but thats not as easy with windows. I have it checking in my git push hook so it tells me when i forget
